### PR TITLE
fix #3802 - Copy array attributes from another array

### LINKED ISSUE #3802: Pick up the array attributes of another array
Would be nice to pick up the array attributes of another array

image

### PR DESCRIPTION
**Pull request by theoryshaw**
https://github.com/IfcOpenShell/IfcOpenShell/assets/507113/7fae2a1e-476a-42ea-8db2-a137b8e288bd

### CODE REVIEW

File: src/blenderbim/blenderbim/bim/module/model/prop.py [hunk 1]
```diff
@@ -87,6 +87,41 @@ def update_type_page(self, context):
...
 
 
+def update_relating_array_from_object(self, context):
+    if self.relating_array_object is None:
+        return
+    element = tool.Ifc.get_entity(self.relating_array_object)
+    if not element:
+        return
+    # If a child is selected, the following assigns the parent to the relating_array_object
+    parent_globalid = element.IsDefinedBy[0].RelatingPropertyDefinition.HasProperties[0].NominalValue.wrappedValue
```
Comment (brunoperdigao) [inline src/blenderbim/blenderbim/bim/module/model/prop.py:97]:
```suggestion
    pset = ifcopenshell.util.element.get_pset(element, "BBIM_Array")
    parent_globalid = pset["Parent"]
```

File: src/blenderbim/blenderbim/bim/module/model/prop.py [hunk 2]
```diff
@@ -87,6 +87,41 @@ def update_type_page(self, context):
...
+    parent_element = tool.Ifc.get().by_guid(parent_globalid)
+    parent_of_relating_array_object = tool.Ifc.get_object(parent_element)
+    #The following error handling is used to break out when an array parent is selected
+    try:
+        element.IsDefinedBy[0].RelatingPropertyDefinition.HasProperties[1].NominalValue.wrappedValue
+    except AttributeError:
```
Comment (brunoperdigao) [inline src/blenderbim/blenderbim/bim/module/model/prop.py:103]:
```suggestion
        pset['Data']
    except Exception:
```

File: src/blenderbim/blenderbim/bim/module/model/prop.py [hunk 3]
```diff
@@ -87,6 +87,41 @@ def update_type_page(self, context):
...
+        pass
+    else:
+        bpy.ops.bim.enable_editing_array(item=self.is_editing)
+        return 
+
+    self.relating_array_object = parent_of_relating_array_object
+    bpy.ops.bim.enable_editing_array(item=self.is_editing)
+
+    return self.relating_array_object
```
Comment (brunoperdigao) [inline src/blenderbim/blenderbim/bim/module/model/prop.py:112]: I was getting an error that the function should return `None`. 
```suggestion
    return
```

File: src/blenderbim/blenderbim/bim/module/model/prop.py [hunk 4]
```diff
@@ -87,6 +87,41 @@ def update_type_page(self, context):
...
+
+
+def is_object_array_applicable(self, obj):
+    element = tool.Ifc.get_entity(obj)
+    if not element:
+        return False
+    return element.IsDefinedBy[0].RelatingPropertyDefinition.Name == "BBIM_Array"
```
Comment (brunoperdigao) [inline src/blenderbim/blenderbim/bim/module/model/prop.py:119]:
```suggestion
    return ifcopenshell.util.element.get_pset(element, "BBIM_Array")
```

File: src/blenderbim/blenderbim/bim/module/model/array.py [hunk 5]
```diff
@@ -83,17 +83,32 @@ class EnableEditingArray(bpy.types.Operator, tool.Ifc.Operator):
...
+            props.method = relating_props.method
+        else:
+            data = json.loads(ifcopenshell.util.element.get_pset(element, "BBIM_Array", "Data"))[self.item]
+            si_conversion = ifcopenshell.util.unit.calculate_unit_scale(tool.Ifc.get())
+            props.x = data["x"] * si_conversion
+            props.y = data["y"] * si_conversion
+            props.z = data["z"] * si_conversion
+            props.use_local_space = data.get("use_local_space", False)
+            props.sync_children = data.get("sync_children", False)
+            props.method = data.get("method", "OFFSET")
```
Comment (brunoperdigao) [inline src/blenderbim/blenderbim/bim/module/model/array.py:108]: You can simplify this:

```suggestion
            element = tool.Ifc.get_entity(relating_obj)
        data = json.loads(ifcopenshell.util.element.get_pset(element, "BBIM_Array", "Data"))[self.item]
        props.count = data["count"]
        si_conversion = ifcopenshell.util.unit.calculate_unit_scale(tool.Ifc.get())
        props.x = data["x"] * si_conversion
        props.y = data["y"] * si_conversion
        props.z = data["z"] * si_conversion
        props.use_local_space = data.get("use_local_space", False)
        props.sync_children = data.get("sync_children", False)
        props.method = data.get("method", "OFFSET")
```

**Review by brunoperdigao — COMMENTED** [review]
It wasn't working properly on my end, so I made a few suggestions that I think we'll make it work and also improve the readability. I'm still getting a warning from Blender "Failed to set value" which I don't fully understand why. However, it doesn't seem to break the functionality

**Review by theoryshaw — COMMENTED** [review]
Thanks!  I'll try these changes.

**Comment by theoryshaw** [comment]
Thanks @brunoperdigao!
The one thing that doesn't seem to work anymore, is that previously you could pick either the parent or the child, and it would work.  Now only picking the parent work.  Picking the child, does not work.

**Review by theoryshaw — COMMENTED** [review]
Thanks!  will try.

**Comment by theoryshaw** [comment]
I think we got it now. :)
much more elegant.

**Comment by brunoperdigao** [comment]
@theoryshaw Nice!
This implementation has one limitation though. In cases where the relating object have more than one array, it will only copy the attributes of the first array. Not a huge problem, but it's something to keep in mind for future improvements.

**Pull request merged**